### PR TITLE
compute: better-fueled linear join

### LINKED ISSUE #19093: Differential join does not yield within keys
### What version of Materialize are you using?

main

### What is the issue?

A current limitation of the DD join implementation (which we use in Materialize to implement differential/linear joins) is that it only checks its remaining fuel in between two processed keys. This means that if there is a lot of work to do for a single key, the operator does not yield until it is done with all that work. Since timely is cooperatively scheduled, this means that now other dataflows get a chance to run for potentially extended amounts of times, and the replica cannot respond to commands from the controller, rendering it unresponsive to user requests.

This unresponsiveness is particularly problematic because it prevents users from querying introspection sources. Without access to introspection sources, it is difficult to diagnose the reason for the unresponsiveness (namely a busy join).

We should fix the fueling logic of the DD join implementation to make it also able to yield within the same key.

### PR DESCRIPTION
**Pull request by teskje**
This PR changes the way linear joins are fueled. Rather than checking the remaining fuel only between keys, the new approach checks every time all matches for a batch value have been produced.

Doing so enables the join operator to regularly yield even when computing joins that have only a single or few keys. This should lead to much better responsiveness of replicas that have dataflows with cross-joins or skewed joins installed.

As this fueling change requires more control over the internals of the join operator than the DD API affords us, part (really: the bulk) of this PR is concerned with copying the join operator code from DD into the MZ codebase. In doing so, it also significantly simplifies the original join operator code, by removing logic and genericity not required by MZ. In general, no attempt is made at keeping the compute join implementation close to the one in DD. This will make it harder to upstream changes, should we ever want to do so, but at the same time gives us the opportunity to make the code more readable. I anticipate that we will want to perform more tuning on the join code in the future (e.g. by adding a shutdown token), so making it easy to modify seems more important than attempting compatibility with DD.

See performance test results here: https://github.com/MaterializeInc/materialize/pull/19186#issuecomment-1540078479

### Proof of Efficacy

To test that this actually has the desired effect of making the join operator more responsive, we can run a test with a cross join:
```sql
CREATE TABLE t (a int);
INSERT INTO t SELECT generate_series(1, 10000);
SELECT sum(t1.a + t2.a * 10000) FROM t t1, t t2;

-- in a second session:
\timing
\o /dev/null
SELECT * FROM mz_internal.mz_dataflow_operators;
SELECT * FROM mz_internal.mz_dataflow_operators;
SELECT * FROM mz_internal.mz_dataflow_operators;
```

On main, this yields the timings on my laptop:
```
Time: 39232.076 ms (00:39.232)
Time: 239.962 ms
Time: 6.080 ms
```

The first query only returns after the cross join has returned (or roughly at the same time). That's because the join operator never yields, so the replica can't process commands while it is running.

On this branch, these timings are printed:
```
Time: 538.751 ms
Time: 872.802 ms
Time: 865.107 ms
```

Queries return promptly even as the cross join is running. The experience is a bit sluggish, but I think this can be solved with a different (e.g. time-based) fueling strategy. I consider that out-of-scope for this PR though.

### Motivation

  * This PR fixes a recognized bug.

Fixes #19093.

### Tips for reviewer

There are two ways to read this PR:

1. Read the full diff. It's not huge, but perhaps overwhelming if you are not familiar with the DD join operator. Most of the positive changes are taken verbatim from the DD implementation and it's not easy to make out what the changes are. So I would generally recommend you go with option (2).
2. Look at the commits separately. I've tried to structure them in a way that makes it obvious what changes are made from the DD join implementation. The first commit simply copies over the relevant code from DD without changes, the subsequent ones modify it in various ways (applying rustfmt, fixing clippy warnings, simplifying). The second-to-last commit introduces the change in fueling. Look at the commit messages for more details.

~~Do you think this change warrants being rolled out with a feature flag? Personally I think that's not necessary, as we don't actually introduce new logic, just modify what's already existing. But I'd be fine with adding a feature flag if folk would like to move slower with this.~~ We agreed on adding a feature flag because there are (unlikely) workloads for this new join implementation is considerably slower than the old one and we want to option of switching back quickly. 

Also let me know if you think I should run other performance tests to ensure there are no regressions!

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR has an associated up-to-date [design doc](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/README.md), is a design doc ([template](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/design/00000000_template.md)), or is sufficiently small to not require a design.
  
- [x] This PR evolves [an existing `$T ⇔ Proto$T` mapping](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/command-and-response-binary-encoding.md) (possibly in a backwards-incompatible way) and therefore is tagged with a `T-proto` label.
- [x] If this PR will require changes to cloud orchestration, there is a companion cloud PR to account for those changes that is tagged with the release-blocker label ([example](https://github.com/MaterializeInc/cloud/pull/5021)).
  
- [x] This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):
  - Make replicas that serve dataflows with cross joins or joins with only few keys stay responsive to user queries.

### CODE REVIEW

**Comment by teskje** [comment]
@philip-stoev This touches a fundamental part of compute, so I think having more eyes on it is warranted. I'm starting a nightly test run now, but lmk if there is anything else we can/should test.

**Comment by teskje** [comment]
Nightlies are green.

The coverage run revealed a couple places that are not covered in the new code, so I added another commit adding tests to fix this. New coverage run: https://buildkite.com/materialize/coverage/builds/96

**Review by vmarcos — COMMENTED** [review]
The changes in the fueling of the join operator look good to me. I have some requests that are more process-based than focused on the change to the code itself:

1. This change takes a stance that we are definitely copying over the join operator from DD, putting in some effort to simplify its code, and not holding on to the hope of upstreaming any changes. In a sense, this is kind of forking off this operator, but without much support to track changes made upstream. This may be fine assuming the upstream operator really does not change much, but I would be more comfortable with this approach if we got an explicit OK from @antiguru  or @frankmcsherry.
2. The experimental data seem to have been obtained with only one worker? It would be ideal if we could repeat the same performance tests proposed here, but under multi-worker setups (with as many workers as feasible). That way, we can contrast the observable overheads, if any.
3. There is not a lot of skew in TPC-H, unfortunately. It would be ideal to test at least a single-query microbenchmark with multiple workers where the join is exposed to skew so as to increase our confidence in this change.

**Comment by teskje** [comment]
Thanks for the review @vmarcos!

1. My assumption is indeed that the DD join operator will not change much in the future. But also a) getting changes upstreamed will be difficult due to Frank's limited bandwidth, b) scoping the implementation to MZ means a significant reduction of complexity, as we can make a lot of assumptions that DD code can't, and c) the extent of the operator is small enough that it seems feasible to keep up with changes in DD even if that involves translating them to the diverged MZ implementation. That said, I appreciate that this might be a controversial stance, so happy to hear other opinions!
2. Sure, I will run the benchmarks with multiple workers/processes too!
3. Would the cross-join suffice as a skewed workload? Or were you thinking of something not _that_ skewed?

**Comment by rjnn** [comment]
I'm signed off on the fork approach, but let's also get explicit sign off from Frank.

File: src/compute/src/render/join/linear_join.rs [hunk 1]
```diff
@@ -313,3 +316,442 @@ where
...
+                            });
+                            batch.step_val(batch_storage);
+                        }
+                        batch.rewind_vals(batch_storage);
+                        trace.step_val(trace_storage);
+
+                        // TODO: This consolidation is optional, and it may not be very
+                        //       helpful. We might try harder to understand whether we
+                        //       should do this work here, or downstream at consumers.
+                        // TODO: Perhaps `thinker` should have the buffer, do smarter
```
Comment (tokenrove) [inline src/compute/src/render/join/linear_join.rs:731]: `thinker` no longer exists, maybe reword this comment?

**Comment by philip-stoev** [comment]
In a branch, I set the fuel constant to 1K vs 1M and kicked-off a Nightly. Please stand by for results ...

File: src/compute/src/render/join/linear_join.rs (hunk 1 above)
Comment (teskje) [reply src/compute/src/render/join/linear_join.rs]: I removed the comment. I don't think with the simple nested-loop matching approach there is not much we can do to consolidate smarter.

**Comment by teskje** [comment]
> I pushed a .td test based on your original manual test that confirms that yielding indeed has the desired end-to-end result. 

This is awesome, thank you!

**Comment by teskje** [comment]
## Performance Tests -- Multi-worker

I ran the same performance tests from above again in a multi-worker setup. Specifically, I made the following changes:

* Use a `medium` replica, rather than an `xsmall`.
* Run TPC-H with scale-factor 10 rather than 1.

These are the results:

### Cross-join

Results [ms]:

main | this branch | ratio
---: | ---: | ---:
77585 | 69070 | 0.89

This is very similar to the single-worker result, which makes sense since a cross-join only has a single key, so pretty much all the processing only happens on a single worker anyway.

### TPC-H

Results [ms]:

query | main | this branch | ratio
:--- | ---: | ---: | ---:
Q01 | 57304 | 53363 | 0.93
Q02 | 7878 | 7997 | 1.02
Q03 | 48020 | 39439 | 0.82
Q04 | 45411 | 41544 | 0.91
Q05 | 47072 | 47225 | 1.0
Q06 | 47719 | 35978 | 0.75
Q07 | 46390 | 48946 | 1.06
Q08 | 52835 | 51301 | 0.97
Q09 | 67074 | 69738 | 1.04
Q10 | 48232 | 50534 | 1.05
Q11 | 6261 | 5927 | 0.95
Q12 | 36923 | 37267 | 1.01
Q13 | 16386 | 15775 | 0.96
Q14 | 39009 | 33778 | 0.87
Q15 | 35124 | 34621 | 0.99
Q16 | 6678 | 7319 | 1.1
Q17 | 52040 | 45285 | 0.87
Q18 | 73589 | 67643 | 0.92
Q19 | 40166 | 38636 | 0.96
Q20 | 375575 | 339703 | 0.9
Q21 | 59490 | 55717 | 0.94
Q22 | 14235 | 12261 | 0.86

This shows a lot more variability than the single-worker case. Measurements generally look to be in favor of the new join implementation, but I wouldn't trust them too much. For example, Q01 doesn't even contain a join and shows 7% difference. I also noticed quite a large variability between benchmark runs on the same MZ build, so I assume the numbers would even out over more runs. Still the measurements seem generally in favor of the new implementation too.

**Comment by teskje** [comment]
## Performance Tests -- Multi-worker -- Skewed Join

And here is the promised test using a skewed join workload.

I created a skewed data distribution using this simple Python code (credits go to @tokenrove):
```python
import random
from collections import Counter

counts = Counter(random.randint(0, 2**64-1).bit_length() for _ in range(20000))
for k, n in enumerate(counts.values()):
    print(f"INSERT INTO t SELECT {k}, generate_series(1, {n});")
```

... then used the output to run this experiment:
```sql
CREATE TABLE t (a int, b int);
INSERT INTO t SELECT 0, generate_series(1, 9924);
INSERT INTO t SELECT 1, generate_series(1, 5064);
INSERT INTO t SELECT 2, generate_series(1, 2557);
INSERT INTO t SELECT 3, generate_series(1, 157);
INSERT INTO t SELECT 4, generate_series(1, 1271);
INSERT INTO t SELECT 5, generate_series(1, 584);
INSERT INTO t SELECT 6, generate_series(1, 73);
INSERT INTO t SELECT 7, generate_series(1, 291);
INSERT INTO t SELECT 8, generate_series(1, 1);
INSERT INTO t SELECT 9, generate_series(1, 43);
INSERT INTO t SELECT 10, generate_series(1, 10);
INSERT INTO t SELECT 11, generate_series(1, 14);
INSERT INTO t SELECT 12, generate_series(1, 8);
INSERT INTO t SELECT 13, generate_series(1, 3);

\timing
SELECT sum(t1.b + t2.b * 10000) FROM t t1 JOIN t t2 USING (a); 
```

The experiment was again run on a `medium`-size replica. This size uses 12 workers, so each worker gets _some_ processing to do.

Results (best of three) [ms]:

<!DOCTYPE html>

| main | this branch | ratio |
| ---: | ---: | ---: |
| 100343 | 78881 | 0.79 |

The performance wins here seem to be more significant than for the cross-join. We should probably again be wary of reading too much into the results considering the variability seen with TPC-H, but I think it's save to say that we don't need to expect regressions.

**Review by vmarcos — APPROVED** [review]
Thanks for the additional evaluation! I am assuming that even if you are running a multi-worker skewed join, you can of course run other queries without losing interactivity. If that is the case and Frank signs off, then no more pending items from my side.

This change is going to make a world of difference for interactivity and debugging when Materialize is exposed to heavy joins.

**Comment by teskje** [comment]
> I am assuming that even if you are running a multi-worker skewed join, you can of course run other queries without losing interactivity.

Yes, these are the query times from the above skewed-join workload:

```sql
materialize=> \timing
Timing is on.
materialize=> \o /dev/null
materialize=> select * from mz_internal.mz_dataflow_operators;
select * from mz_internal.mz_dataflow_operators;
select * from mz_internal.mz_dataflow_operators;
Time: 1225.766 ms (00:01.226)
Time: 1443.466 ms (00:01.443)
Time: 1446.125 ms (00:01.446)
```

It's not snappy but still a huge improvement over potentially infinite query times :)

**Comment by teskje** [comment]
@antiguru pointed out a stray `#[inline(never)]` in the join code that probably doesn't make sense to have. I removed it and it turns out to be great for performance!

These are comparisons between the second-to-last commit (with disallowed inlining) and the last commit, using the performance tests from above:

### Cross-join

before | now | ratio
---: | ---: | ---:
69070 | 66629 | 0.96 

### Skewed join

before | now | ratio
---: | ---: | ---:
78881 | 68986 | 0.87 

### TPC-H

query | before | now | ration
:--- | ---: | ---: | ---:
Q01 | 66974 | 53363 | 0.80
Q02 | 13377 | 7997 | 0.60
Q03 | 46605 | 39439 | 0.85
Q04 | 49483 | 41544 | 0.84
Q05 | 57777 | 45670 | 0.79
Q06 | 39902 | 35978 | 0.90
Q07 | 61234 | 48946 | 0.80
Q08 | 58986 | 51301 | 0.87
Q09 | 68018 | 69738 | 1.03
Q10 | 69203 | 50534 | 0.73
Q11 | 11340 | 5927 | 0.52
Q12 | 52928 | 37267 | 0.70
Q13 | 17936 | 15775 | 0.88
Q14 | 37374 | 33485 | 0.90
Q15 | 35227 | 34621 | 0.98
Q16 | 7073 | 7319 | 1.03
Q17 | 58866 | 45285 | 0.77
Q18 | 75762 | 66622 | 0.88
Q19 | 47748 | 37520 | 0.79
Q20 | 339554 | 326566 | 0.96
Q21 | 107615 | 55717 | 0.52
Q22 | 14880 | 12251 | 0.82

As always, caveats apply due to the high variance of especially the TPC-H results.

**Comment by teskje** [comment]
# Worst-case Performance Test

Here is a performance test that favors the merge join strategy that this PR removes from the linear join. The test generates 1000 events every second, one event at each millisecond. An event consists of an insert and a retraction, with the retraction removing the value from the previous insert.

Timely batches look something like this:
```
at time 0000: [(0, 0, +1), (0, 1, -1), (1, 1, +1), (1, 2, -1), (2, 2, +1), ...]
at time 1000: [(999, 1000, -1), (1000, 1000, +1), (1000, 1001, -1), (1001, 1001, +1), ...]
...  
```

Basically, we count up at the smallest granularity possible in Materialize, retracting and inserting a value every ms. The frontier ticks up once every second. This gives the join batches of 2000 updates with different times. Having different times favors the merge join strategy, which doesn't have any advantage over the nested loop join if all updates in a batch have the same time.

## Test Setup

It's not easy to generate the above data in Materialize today. Source rechecking works with second-granularity, and rounds all updates to the next second-boundary. @petrosagg points out that CDCv2 envelopes can be used to inject arbitrary timestamps, but I didn't want to mess around with Kafka. Instead I patched the compute rendering code like this:

```diff
diff --git a/src/compute/src/render/mod.rs b/src/compute/src/render/mod.rs
index 1d1294c2cf..9c88505247 100644
--- a/src/compute/src/render/mod.rs
+++ b/src/compute/src/render/mod.rs
@@ -110,9 +110,11 @@ use differential_dataflow::operators::arrange::Arrange;
 use differential_dataflow::operators::reduce::ReduceCore;
 use differential_dataflow::AsCollection;
 use timely::communication::Allocate;
+use timely::dataflow::channels::pact::Pipeline;
+use timely::dataflow::operators::Operator;
 use timely::dataflow::operators::to_stream::ToStream;
 use timely::dataflow::scopes::Child;
-use timely::dataflow::Scope;
+use timely::dataflow::{Scope, Stream};
 use timely::order::Product;
 use timely::progress::timestamp::Refines;
 use timely::progress::Timestamp;
@@ -233,6 +235,10 @@ pub fn build_compute_dataflow<A: Allocate>(
                         |_timer, count| count > 1_000_000,
                     );

+                    if dataflow.debug_name.contains("join_test") {
+                        ok_stream = gen_join_test_updates(ok_stream);
+                    }
+
                     // If `mfp` is non-identity, we need to apply what remains.
                     // For the moment, assert that it is either trivial or `None`.
                     assert!(mfp.map(|x| x.is_identity()).unwrap_or(true));
@@ -389,6 +395,48 @@ pub fn build_compute_dataflow<A: Allocate>(
     })
 }

+fn gen_join_test_updates<G>(
+    stream: Stream<G, (Row, mz_repr::Timestamp, mz_repr::Diff)>,
+) -> Stream<G, (Row, mz_repr::Timestamp, mz_repr::Diff)>
+where
+    G: Scope<Timestamp = mz_repr::Timestamp>,
+{
+    const UPDATES_PER_TICK: u64 = 1000;
+
+    stream.unary(Pipeline, "gen_join_test_updates", |_cap, _info| {
+        let mut scratch = Vec::new();
+        let mut to_retract = None;
+        move |input, output| {
+            input.for_each(|cap, data| {
+                data.swap(&mut scratch);
+                let time = cap.time().clone();
+                let mut session = output.session(&cap);
+
+                for (row, _, diff) in scratch.drain(..) {
+                    assert_eq!(diff, 1);
+                    let mz_repr::Datum::Int64(tick) = row.unpack_first() else {
+                        panic!("unexpected row format: {row:?}");
+                    };
+
+                    let start_idx = (tick as u64) * UPDATES_PER_TICK;
+                    for i in 0..UPDATES_PER_TICK {
+                        let idx = start_idx + i;
+                        let t = time.step_forward_by(&i.into());
+
+                        if let Some(row) = to_retract.take() {
+                            session.give((row, t, -1));
+                        }
+
+                        let row = Row::pack([mz_repr::Datum::Int64(idx as i64)]);
+                        to_retract = Some(row.clone());
+                        session.give((row, t, 1));
+                    }
+                }
+            });
+        }
+    })
+}
```

This adds an additional operator behind every source in dataflows containing the magic name `join_test`. That operator expects to be installed after a `COUNTER` load generator that ticks once every second. It modifies the tick updates into the format described above.

With this patch applied, we create two `join_test` dataflows and feed their outputs into a join:
```sql
CREATE SOURCE tick FROM LOAD GENERATOR COUNTER (TICK INTERVAL '1s') WITH (SIZE '1');

CREATE CLUSTER producer REPLICAS (r (SIZE '1'));
SET CLUSTER = producer;

CREATE MATERIALIZED VIEW join_test_1 AS SELECT * FROM tick;
CREATE MATERIALIZED VIEW join_test_2 AS SELECT * FROM tick;

CREATE CLUSTER consumer REPLICAS (r (SIZE '1'));
SET CLUSTER = consumer;

CREATE VIEW join AS SELECT a.counter a, b.counter b FROM join_test_1 a, join_test_2 b;
CREATE DEFAULT INDEX ON join;
```

## Results

TODO!

**Review by vmarcos — APPROVED** [review]
Updates LGTM!

File: src/compute/src/render/join/mod.rs [hunk 2]
```diff
@@ -13,3 +13,5 @@
 
 mod delta_join;
 mod linear_join;
+
+pub use linear_join::JoinImpl as LinearJoinImpl;
```
Comment (antiguru) [inline src/compute/src/render/join/mod.rs:17]: Why not just name `JoinImpl` `LinearJoinImpl`? This would remove one level of indirection.

File: src/sql/src/session/vars.rs [hunk 3]
```diff
@@ -874,6 +874,17 @@ pub const ENABLE_SESSION_RBAC_CHECKS: ServerVar<bool> = ServerVar {
     internal: false,
 };
 
+/// Whether compute rendering should use the new linear join implementation rather than the one
+/// from Differential Dataflow.
+const ENABLE_NEW_LINEAR_JOIN_IMPLEMENTATION: ServerVar<bool> = ServerVar {
```
Comment (antiguru) [inline src/sql/src/session/vars.rs:879]: A nit, and not to be solved in this PR. We should figure out better names for our feature flags. `NEW_LINEAR_JOIN` does not convey anything about the nature of the change, and ideally it should be clear from the description what it does. Something that's new today will be old tomorrow :D

File: src/compute/src/render/join/linear_join.rs [hunk 4]
```diff
@@ -312,3 +370,435 @@ where
         (oks, None)
     }
 }
+
+/// Joins two arranged collections with the same key type.
+///
+/// Each matching pair of records `(key, val1)` and `(key, val2)` are subjected to the `result` function,
+/// which produces something implementing `IntoIterator`, where the output collection will have an entry for
+/// every value returned by the iterator.
+fn mz_join_core<G, Tr1, Tr2, L, I>(
```
Comment (antiguru) [inline src/compute/src/render/join/linear_join.rs:379]: Maybe it's worth moving this to a separate module or file, so we can have a higher-level comment explaining why we're adding our own version of `join_core`, and to keep it more separate from uses.

**Review by antiguru — APPROVED** [review]

File: src/sql/src/session/vars.rs (hunk 3 above)
Comment (teskje) [reply src/sql/src/session/vars.rs:879]: Well, the idea is to arrive at a single join implementation and remove that flag before the new one becomes old :) But I also don't like the name. Ideally we'd have one that conveys that a) the new join has better yielding behavior than the DD join and b) the new join is inferior for workloads with fine-grained times. But I can't come up with anything good.

We could also side-step the issue, invert the meaning of the flag and call it `enable_dd_linear_join_implementation`. But then we'd have a feature flag that needs to be set to `false` to enable the feature it protects, which would be confusing.

File: src/compute/src/render/join/linear_join.rs (hunk 4 above)
Comment (teskje) [reply src/compute/src/render/join/linear_join.rs:379]: Good idea!

File: src/sql/src/session/vars.rs (hunk 3 above)
Comment (vmarcos) [reply src/sql/src/session/vars.rs:879]: How about `enable_mz_join_core`?

Comment (teskje) [reply src/sql/src/session/vars.rs]: Thanks, that seems like a better name. Changed.

**Comment by teskje** [comment]
TFTRs!
